### PR TITLE
Master

### PR DESCRIPTION
**Pull request by ericlippert**
* Hack reserves keywords "parent" and "self"; this was not documented
in the specification. (Interestingly, Hack does not reserve "null",
"true", "false" or "this", which we might want to look into.)

* The explanation of the ${ } interpolation in string literals had a
link to a now-deleted section of the specification, which meant that
the feature was not explained. Since this feature is very restricted
compared to its equivalent in PHP -- the PHP version of it cannot even
be *lexed* without direction from the parser! -- I've added some
verbiage explaining the difference and giving examples.

* The deleted feature mentioned in the point above is the Variable-Name
creation operator, but does not give the syntax which has been removed.
The ${$y} syntax is at present unsupported in Hack. The $$y syntax is
supported incorrectly now; Alex has a diff outstanding to make it an
error.

* The link to the "enum declarations" link was copy-pasted wrong.

* Text and links for the require-once directive mistakenly characterized it as an expression in some places

* Based on feedback I've removed the sections on the ${} string interpolation entirely. They are invalid in strict mode and hence not part of this specification.

* Fixed a typo in the grammar for return-type.

* Fix some broken cross-references

### CODE REVIEW

File: spec/09-lexical-structure.md [hunk 1]
```diff
@@ -678,7 +678,7 @@ The variable substitution accepts the following syntax:
 * [*name*](#names)
 * [*variable-name*](#names)
 
-*expression* works the same way as in [variable name creation operator](10-expressions.md#variable-name-creation-operator).
+The `${ expression }` syntax is deprecated in Hack. The expression must be such that if the enclosing braces were removed and the `$` prepended, then the resulting expression would be a legal <i>`string-variable`</i>. For example, `"${y[123]}"` is legal because `$y[123]` is a legal <i>`string-variable`</i>. No whitespace or comments are allowed inside the braces.
```
Comment (JoelMarcey) [inline spec/09-lexical-structure.md]: Got some feedback from @RexJaeschke ....

> As a spec writer, I am VERY much opposed to using the terms “legal” and “illegal” when the subject matter has nothing to do with the law. I propose changing these to “valid” and “invalid”, respectively, or “well formed” and “ill formed”.

> Re “The `${ expression }` syntax is deprecated in Hack.”, in the context of a standard, “deprecated” means “supported now but might not be in future versions”. If this construct truly is NOT supported by Hack now, then “deprecation” is not the right word to use here.

File: spec/23-differences-from-php.md [hunk 2]
```diff
@@ -135,6 +135,12 @@ This is not supported.
 ###Logical `AND`, `OR`, `XOR` Operators (Alternate Forms)
 The `and`, `xor`, and `or` alternate forms are not supported.
 
+###String Literals
+
+PHP allows the syntax `"${ expression }"` in a double-quoted string literal, where the expression can be the name of a variable, or any expression which evaluates to a string naming a variable. In PHP the expression may include delimited comments, spaces, and so on. In Hack this syntax is both deprecated and more strict. In Hack the only expression that may appear inside the braces is an expression which would be a legal <i>`string-variable`</i> if the braces were removed. 
+
+For example, `"${y[123]}"` would be legal in Hack because `$y[123]` is a legal <i>`string-variable`</i>. In contrast, `"${ 'x'}"` is legal PHP equivalent to "$x", but in Hack this is not legal because `$ 'x'` is not a legal <i>`string-variable`</i>.
```
Comment (JoelMarcey) [inline spec/23-differences-from-php.md]: See `legal` comment above.

File: spec/09-lexical-structure.md (hunk 1 above)
Comment (ericlippert) [inline spec/09-lexical-structure.md]: Great feedback, thanks. 

Couple things.

First, I totally agree about legal vs valid. I'll make those fixes.

Second, I did not realize when I wrote this that the specification says in section 4 that the specification is only specifying "strict" mode. What I intended to convey by the comment about deprecation is that this syntactic form produces an error in strict mode. 

So I'm not sure what to do at this point. We could (1) remove it entirely and be done with it, or (2) add a note here saying that this syntax is permitted in non-strict mode, or (3) remove it from here, make a new appendix that is "common deviations from strict mode", (4) leave the text as-is and not stress about it, or (5) do something I haven't thought of.  Thoughts and opinions?

File: spec/23-differences-from-php.md (hunk 2 above)
Comment (ericlippert) [inline spec/23-differences-from-php.md]: Will fix.  Thanks!

File: spec/09-lexical-structure.md (hunk 1 above)
Comment (RexJaeschke) [inline spec/09-lexical-structure.md]: Clause 4 states, “This specification is written from the perspective of strict mode only. A conforming implementation may provide modes other than strict, but they are outside the scope of this specification.”
 
Once we established that we'd limit things to strict, I believe I didn’t make any other references to non-strict modes, so to be consistent, we probably should continue to say nothing about them. If we do say something here, then that opens up the sticky question of “what else could/should we say in other places about other things?".

Comment (ericlippert) [inline spec/09-lexical-structure.md]: I agree. I'll delete the text for the not-supported-in-strict-mode feature entirely.

I am starting an internal document describing the differences between strict and lax modes; if we want to turn that into a spec in the future, but for now I'll just use it for my own purposes.

**Pull request merged**